### DISCUSSION TITLE
Android 13 related issue

### DISCUSSION THREAD

**Discussion by Amar1312** [yasirkula/UnityNativeGallery#299] [Q&A] [answered]
Is there any solution for android 13 , working with older version but not with Android 13 because of new permission rule 
<uses-permission android:name="android.permission.READ_MEDIA_IMAGES" />
<uses-permission android:name="android.permission.READ_MEDIA_AUDIO" />

**Comment by yasirkula** [comment, marked as answer]
Can you update the plugin by downloading it from GitHub, if you haven't already done so?

**Reply by Amar1312** [reply]
Thanks solved it by updating

**Comment by IgorCFG** [comment]
@yasirkula i still have this issue and already update the NativeGallery to 1.8.1, can you help me?

I create this method and it always return true.

```C#
        var permission = await NativeGallery.RequestPermissionAsync(
            NativeGallery.PermissionType.Read | NativeGallery.PermissionType.Write, 
            NativeGallery.MediaType.Image);

        return permission == NativeGallery.Permission.Granted;
```

Android popup that ask for permission never show and the error on logcat still that:
2024/10/09 15:56:27.270 31863 32453 Error Unity java.io.FileNotFoundException: /storage/emulated/0/Android/media/com.whatsapp/WhatsApp/Media/WhatsApp Images/IMG-20241005-WA0007.jpeg: open failed: EACCES (Permission denied)

**Reply by yasirkula** [reply]
It's normal for the permission dialog to not show up. Can you share the rest of your script? Maybe I can spot an issue.

**Reply by IgorCFG** [reply]
Sure.

I was testing in Bluestacks device and works fine.
I even wondered if it was some security lock on the Xiaomi system, which has some pretty annoying things.

```cs


    public async Task<byte[]> PickFromGalleryAsync()
    {
        var isConfirmed = await _popupManager.ShowPopupWithConfirmationAsync(
            "{{permission_title}}",
            "{{image_permission_message}}"
        );

        if (!isConfirmed) return null;

        // var hasPermission = await RequestPermissionAsync();
        // if (!hasPermission) return null;

        var tcs = new TaskCompletionSource<byte[]>();

        _ = NativeGallery.GetImageFromGallery((path) =>
        {
            if (path != null)
            {
                var texture = NativeGallery.LoadImageAtPath(
                    path,
                    markTextureNonReadable: false);

                if (texture != null && texture.isReadable) 
                {
                    var resizedTexture = ResizeImage(texture, 128, 128);
                    var imageData = resizedTexture.EncodeToPNG();

                    tcs.SetResult(imageData);
                }
            }

            tcs.SetResult(null);
        }, "Selecione uma imagem", "image/*");

        return await tcs.Task;
    }
    
    private async Task<bool> RequestPermissionAsync()
    {
        var permission = await NativeGallery.RequestPermissionAsync(
            NativeGallery.PermissionType.Read | NativeGallery.PermissionType.Write, 
            NativeGallery.MediaType.Image);

        return permission == NativeGallery.Permission.Granted;
    }

```

**Reply by IgorCFG** [reply]
Check this @yasirkula 
![image](https://github.com/user-attachments/assets/15886344-1db5-4aa7-ad6b-826d22fb7e6f)

I get the error again when I remove the `markTextureNonReadable: false`.
I was not aware of this parameter until I opened NativeGallery source code.

It's weird.. The Exception was `FileNotFoundException` and I using the same file as before, my guess is that your library recreates the file in another folder and without this parameter it doesn't seem to recreate correctly to Texture2D right? But even so, the exception left me very confused when debugging.

I using that script to Resize the image too, so probably is that, sorry for that.
```cs

    public Texture2D ResizeImage(Texture2D sourceTexture, int maxWidth, int maxHeight)
    {
        float widthRatio = (float)maxWidth / sourceTexture.width;
        float heightRatio = (float)maxHeight / sourceTexture.height;
        float resizeRatio = Mathf.Min(widthRatio, heightRatio);

        int newWidth = Mathf.RoundToInt(sourceTexture.width * resizeRatio);
        int newHeight = Mathf.RoundToInt(sourceTexture.height * resizeRatio);

        var resizedTexture = new Texture2D(newWidth, newHeight, sourceTexture.format, false);
    
        var pixelData = sourceTexture.GetPixels(0, 0, sourceTexture.width, sourceTexture.height);
        var resizedData = ResizePixelArray(pixelData, sourceTexture.width, sourceTexture.height, newWidth, newHeight);

        resizedTexture.SetPixels32(resizedData);
        resizedTexture.Apply();

        return resizedTexture;
    }

    private Color32[] ResizePixelArray(Color[] originalPixels, int originalWidth, int originalHeight, int newWidth, int newHeight)
    {
        var resizedPixels = new Color32[newWidth * newHeight];

        for (int y = 0; y < newHeight; y++)
        {
            for (int x = 0; x < newWidth; x++)
            {
                float xRatio = (float)x / newWidth;
                float yRatio = (float)y / newHeight;

                int originalX = Mathf.RoundToInt(xRatio * originalWidth);
                int originalY = Mathf.RoundToInt(yRatio * originalHeight);

                resizedPixels[x + y * newWidth] = originalPixels[originalX + originalY * originalWidth];
            }
        }

        return resizedPixels;
    }
```

**Reply by yasirkula** [reply]
Oh, that's actually a harmless log that's coming from here: https://github.com/yasirkula/UnityNativeGallery/blob/4d27475bed66d56aca8b7d585b98fe503525e593/.github/AAR%20Source%20(Android)/java/com/yasirkula/unity/NativeGalleryMediaPickerResultOperation.java#L148

It's logged when gallery returns an inaccessible path. When that happens, NativeGallery copies the source file to temporaryCachePath and returns it. Does NativeGallery work as intended if you ignore this error in logcat? I was assuming that the texture wasn't being loaded.

markTextureNonReadable should be a coincidence. Looking at the source code, I can't spot any lines where it could cause issues.

**Reply by IgorCFG** [reply]
Hi @yasirkula!

If I use the Resize methods it don't work without `markTextureNonReadable = false`.
I can try to use directly the Texture2D generated from `LoadImageAtPath` and set as sprite in my Image, but only for test purpose, do you need this test?

**Reply by yasirkula** [reply]
markTextureNonReadable does have effect on EncodeToPNG or GetPixels' functionality. It must be false for those to work. If not, a UnityException will be thrown. I was focused on FileNotFoundException in my previous message and missed this case. If the only issue with markTextureNonReadable is UnityException, I don't need you to test it.

Note that LoadImageAtPath's maxSize parameter is a powerful feature that downscales images if necessary and should be faster and more optimized than hand-written resize functions. But it doesn't support specifying different maxWidth maxHeight values.

**Reply by IgorCFG** [reply]
> I was focused on FileNotFoundException in my previous message and missed this case.

Yeah, I do the same because of that exception message. 😂

> Note that LoadImageAtPath's maxSize parameter is a powerful feature that downscales images if necessary and should be faster and more optimized than hand-written resize functions. But it doesn't support specifying different maxWidth maxHeight values.

Oh, it's nice to known. If i use `maxSize: 128` I'll have the same behavior from my "chat gpt" method 😂? 

I thought it was a filter to only search for images "up to" that size.

**Reply by yasirkula** [reply]
`maxSize: 128` will resize 256x512 image to 64x128 and 512x256 image to 128x64. Resizing is done via native code.
